### DISCUSSION TITLE
Response Converter

### DISCUSSION THREAD

**Discussion by joseluisgs** [Foso/Ktorfit#314] [Q&A]
Hi @Foso 

I will try to make a Response Converter using a Result class. I will try th use the Result Kotlin class https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/

But when I use: 
@GET("api/users/{id}")
    suspend fun getById(@Path("id") id: Int): Result<GetByIdDto>

I will I have
Exception in thread "main" kotlinx.serialization.SerializationException: Serializer for class 'Result' is not found.
Mark the class as @Serializable or provide the serializer explicitly.

This is my implementation

class KtorFitResponseConverter : SuspendResponseConverter {

    override fun supportedType(typeData: TypeData, isSuspend: Boolean): Boolean {
        return typeData.qualifiedName == "com.example.model.MyOwnResponse"
    }

    override suspend fun <RequestType> wrapSuspendResponse(
        typeData: TypeData,
        requestFunction: suspend () -> Pair<TypeInfo, HttpResponse>,
        ktorfit: Ktorfit
    ): Any {
        return try {
            val (info, response) = requestFunction()
            Result.success<Any>(response.body(info))

        } catch (ex: Throwable) {
            Result.failure<Any>(ex)
        }
    }
}

I try to have my own ResponseConverter using your web
https://foso.github.io/Ktorfit/responseconverter/

But I always have problem with serialization

Please could you give me a example or advice to make it using Result Kotlin, my own result or for example https://github.com/michaelbull/kotlin-result

I will appreciate it very much
Than you very much!

**Comment by Foso** [comment]
Hi @joseluisgs , you need to replace **"com.example.model.MyOwnResponse"** with **kotlin.Result** . Otherwise no converter for Result can be found, and Ktor tries to convert the Response to Result

**Comment by joseluisgs** [comment]
Hi @Foso 
Thank you very much. It is great to use it with Result and Flows. You are the great. Ktorfit is my favorite!!!

I include the solution for everyone!!

Another question, my final configuration is ok? Anything else to improve it?

Thank you so much!

```kotlin
class KotlinResultConverter : SuspendResponseConverter {
    override fun supportedType(typeData: TypeData, isSuspend: Boolean): Boolean {
        return typeData.qualifiedName == "kotlin.Result"
    }

    override suspend fun <RequestType> wrapSuspendResponse(
        typeData: TypeData,
        requestFunction: suspend () -> Pair<TypeInfo, HttpResponse>,
        ktorfit: Ktorfit
    ): Any {
        return try {
            val (info, response) = requestFunction()
            Result.success<Any>(response.body(info))
        } catch (ex: Throwable) {
            Result.failure(ex)
        }
    }
}
```
```kotlin
Ktorfit.Builder()
            .httpClient {
                install(ContentNegotiation) {
                    json(Json { isLenient = true; ignoreUnknownKeys = true })
                }
                install(DefaultRequest) {
                    header(HttpHeaders.ContentType, ContentType.Application.Json)
                }
            }
            .baseUrl(API_URL)
            .responseConverter(FlowResponseConverter()) 
            .responseConverter(KotlinResultConverter()) 
            .build().create<KtorFitRest>()
```
